### PR TITLE
fix(posthog): gate proxied requests on trusted egress

### PR DESCRIPTION
**Pull request by yiboyasss**
## Summary
- An ambient HTTP(S)/OS-native proxy performs its own DNS resolution, silently bypassing `_base_url()`'s private-network validation of `POSTHOG_HOST` -- the same class of SSRF bypass just closed in `magento.py`'s `_request()`.
- Gate on `get_trusted_proxy_url()` and issue the request on a `trust_env=False` `requests.Session` so no ambient env-var or OS-native proxy source is picked up unless `XAGENT_TRUSTED_EGRESS_PROXY=1` is explicitly set.

## Test plan
- [x] `tests/web/tools/test_posthog_mcp.py` (existing + new proxy-trust/`trust_env` tests, mirroring `test_magento_mcp.py`)
- [x] `tests/web/tools/` full directory
- [x] `ruff check` / `ruff format`
- [x] `mypy src/xagent/web/tools/mcp/posthog.py`
- [x] Self-review pass (code-review skill, high effort) -- no findings

### CODE REVIEW

**Comment by rogercloud** [comment]
CI is failing on this PR — Pytest Fast (web) is failing. Please provide a green build before requesting review.

File: tests/web/tools/test_posthog_mcp.py [hunk 1]
```diff
@@ -346,11 +346,92 @@ def test_request_uses_configured_host_and_headers(monkeypatch):
...
+
+    class FakeSession:
+        def __enter__(self):
+            self.trust_env = True
+            return self
+
+        def __exit__(self, *exc_info):
+            return False
+
+        def request(self, **kwargs):
```
Comment (rogercloud) [inline tests/web/tools/test_posthog_mcp.py]: **Major, non-blocking.** `FakeSession.request(self, **kwargs)` only captures `trust_env` and ignores every other kwarg. The only test that exercises the real `_make_request` body (`test_make_request_disables_trust_env`) only asserts `trust_env is False`; all other tests that assert on `proxies=`/`allow_redirects=` mock `_make_request` itself away. If `_make_request`'s body silently dropped `proxies=`/`allow_redirects=False` before calling `session.request()` -- stripping the SSRF-egress gate and the redirect/credential-leak guard this PR adds -- the whole suite would still pass.

Suggestion: capture the full `kwargs` dict here and assert `proxies`/`allow_redirects` in `test_make_request_disables_trust_env`.

Blocking: no.

File: src/xagent/web/tools/mcp/posthog.py [hunk 2]
```diff
@@ -190,23 +191,64 @@ def _extract_error_detail(response: requests.Response) -> str | None:
...
+    once no env var is set -- the same DNS-rebinding-via-proxy bypass this
+    gate exists to close, one layer deeper than an env-var scrub alone
+    reaches. Mirrors magento.py's identical `_make_request`.
+
+    `trust_env=False` also disables `requests`' `REQUESTS_CA_BUNDLE`/
+    `CURL_CA_BUNDLE` lookup and `.netrc` auto-auth (both gated behind the
+    same `if self.trust_env:` check in `requests`' own Session code); no
+    re-application is needed here, unlike magento.py's self-hosted target,
+    since PostHog Cloud's two hosts always terminate TLS with a publicly
+    trusted CA and this connector always sends its own Bearer header.
```
Comment (rogercloud) [inline src/xagent/web/tools/mcp/posthog.py]: **Medium, non-blocking.** This docstring's CA-bundle reasoning doesn't hold for the deployment scenario this PR enables: `trust_env=False` also disables `requests`' `REQUESTS_CA_BUNDLE`/`CURL_CA_BUNDLE` lookup (gated behind the same `if self.trust_env:` check), and the only case where a proxy is actually used here is when an operator sets `XAGENT_TRUSTED_EGRESS_PROXY=1` -- a corporate egress proxy, the textbook case for TLS interception with an internally-issued CA, not a publicly trusted one. An operator behind such a proxy who opts in would expect `REQUESTS_CA_BUNDLE` to be honored but instead gets `SSLError: certificate verify failed` on every call, with no way to fix it. Fails closed (no security hole), but the "trusted proxy" feature is non-functional in the deployment most likely to need it.

Suggestion: explicitly apply `REQUESTS_CA_BUNDLE`/`CURL_CA_BUNDLE` to `verify=` when `trust_env=False`, or correct the docstring.

Blocking: no.

File: src/xagent/web/tools/mcp/posthog.py [hunk 3]
```diff
@@ -190,23 +191,64 @@ def _extract_error_detail(response: requests.Response) -> str | None:
...
+    # get_trusted_proxy_url() raises unless the proxy is explicitly marked
+    # trusted to enforce its own private-range egress policy
+    # (XAGENT_TRUSTED_EGRESS_PROXY=1), rather than silently trusting every
+    # ambient proxy setup_proxy_env() promotes from the OS's own settings.
+    # The result is then passed to `_make_request()` (see above), which
+    # issues the call on a trust_env=False Session -- disabling every
+    # ambient/OS-native proxy source, not just the env vars this function
+    # checks -- so "no proxy" here is an actual guarantee for this call.
+    proxy_url = get_trusted_proxy_url()
+    proxies = {"http": proxy_url, "https": proxy_url} if proxy_url else {}
```
Comment (rogercloud) [inline src/xagent/web/tools/mcp/posthog.py:257]: **Minor, non-blocking.** `NO_PROXY` is no longer honored -- a genuine regression vs. pre-PR behavior. `NO_PROXY` bypass logic only runs inside `merge_environment_settings`'s `if self.trust_env:` branch, so with `trust_env=False` plus an explicitly-passed `proxies=` dict it's never consulted. Pre-PR, `requests.request()` ran with default `trust_env=True` and no explicit `proxies=`, so `NO_PROXY=us.posthog.com` was respected; post-PR it isn't once `XAGENT_TRUSTED_EGRESS_PROXY=1` is set. Direction is fail-secure (more traffic through the operator's own already-trusted proxy, not less), so this is an operational note rather than a security issue.

Blocking: no.

File: src/xagent/web/tools/mcp/posthog.py [hunk 4]
```diff
@@ -190,23 +191,64 @@ def _extract_error_detail(response: requests.Response) -> str | None:
...
+    # web_content.py's fetch_web_content and magento.py's _request:
+    # get_trusted_proxy_url() raises unless the proxy is explicitly marked
+    # trusted to enforce its own private-range egress policy
+    # (XAGENT_TRUSTED_EGRESS_PROXY=1), rather than silently trusting every
+    # ambient proxy setup_proxy_env() promotes from the OS's own settings.
+    # The result is then passed to `_make_request()` (see above), which
+    # issues the call on a trust_env=False Session -- disabling every
+    # ambient/OS-native proxy source, not just the env vars this function
+    # checks -- so "no proxy" here is an actual guarantee for this call.
+    proxy_url = get_trusted_proxy_url()
```
Comment (rogercloud) [inline src/xagent/web/tools/mcp/posthog.py]: **Minor, non-blocking.** The documented `XAGENT_TRUSTED_EGRESS_PROXY=1` remedy in this error isn't reachable via the self-service connect flow -- only an admin-level global MCP server config edit works. The self-service `/apps/{app_id}/connect` endpoint only forwards `POSTHOG_API_KEY`/`POSTHOG_HOST` into this connector's subprocess env (`required_env` filter in `src/xagent/web/api/mcp.py`); `PUT /servers/{id}` (admin path, applies `config.env` without that filter) does work. The error message doesn't distinguish "set it as a host/shell env var" (won't work for this connector) from "have an admin add it to the connector's global MCP server config" (does work).

Suggestion: clarify the error message/docs to point at the actual admin remedy.

Blocking: no.

File: src/xagent/web/tools/mcp/posthog.py [hunk 5]
```diff
@@ -190,23 +191,64 @@ def _extract_error_detail(response: requests.Response) -> str | None:
...
+    """Issue the actual HTTP call with `trust_env=False`, so `requests`
+    never falls back to an ambient/OS-native proxy source
+    `get_trusted_proxy_url()` doesn't police -- passing `proxies=` alone
+    isn't enough, since a `trust_env=True` Session (`requests.request()`'s
+    default) still calls `urllib.request.getproxies()`, which itself falls
+    back *past* `HTTP_PROXY`/`HTTPS_PROXY`/`ALL_PROXY` to the OS's own proxy
+    configuration (`getproxies_macosx_sysconf()`/`getproxies_registry()`)
+    once no env var is set -- the same DNS-rebinding-via-proxy bypass this
+    gate exists to close, one layer deeper than an env-var scrub alone
+    reaches. Mirrors magento.py's identical `_make_request`.
```
Comment (rogercloud) [inline src/xagent/web/tools/mcp/posthog.py]: **Minor, non-blocking.** This comment (and lines 209, 229) reference `magento.py`, which doesn't exist in this PR's branch or on `main` -- it exists on a separate unmerged sibling branch by the same author with a near-identical `_make_request`. Accurate today, but a reader of `main` after this PR merges (before, or unless, that sibling PR also lands) will hit a comment pointing at a file they can't find, with no cross-reference in this PR's description.

Suggestion: reference the sibling PR number explicitly, or describe the rationale standalone without naming a not-yet-merged file.

Blocking: no.

File: src/xagent/web/tools/mcp/posthog.py [hunk 6]
```diff
@@ -190,23 +191,64 @@ def _extract_error_detail(response: requests.Response) -> str | None:
...
         for attempt in (0, 1):
-            response = requests.request(
+            response = _make_request(
                 method=method,
                 url=url,
                 headers=_headers(),
                 params=params,
                 json=json_data,
                 timeout=DEFAULT_TIMEOUT_SECONDS,
+                proxies=proxies,
```
Comment (rogercloud) [inline src/xagent/web/tools/mcp/posthog.py:269]: **Minor, non-blocking.** Further down in this function, the `except requests.RequestException` handler's comment (around line 270-274, outside this diff hunk) is now stale: it still says a `ProxyError` could echo "the ambient HTTPS_PROXY URL... setup_proxy_env() exports whatever the OS has configured," but post-PR `_make_request` is only ever called with a proxy that already passed `get_trusted_proxy_url()`'s trust check, or no proxy at all -- an untrusted ambient proxy now raises before reaching this code. The corresponding test comment (`tests/web/tools/test_posthog_mcp.py:493-495`) was updated to say "connecting through a trusted proxy"; this source comment (posthog.py:270-274) was missed.

Blocking: no.

File: src/xagent/web/tools/mcp/posthog.py [hunk 7]
```diff
@@ -190,23 +191,64 @@ def _extract_error_detail(response: requests.Response) -> str | None:
...
+    # (XAGENT_TRUSTED_EGRESS_PROXY=1), rather than silently trusting every
+    # ambient proxy setup_proxy_env() promotes from the OS's own settings.
+    # The result is then passed to `_make_request()` (see above), which
+    # issues the call on a trust_env=False Session -- disabling every
+    # ambient/OS-native proxy source, not just the env vars this function
+    # checks -- so "no proxy" here is an actual guarantee for this call.
+    proxy_url = get_trusted_proxy_url()
+    proxies = {"http": proxy_url, "https": proxy_url} if proxy_url else {}
+
     url = f"{_base_url()}{path}"
```
Comment (rogercloud) [inline src/xagent/web/tools/mcp/posthog.py:259]: **Minor, non-blocking.** The proxy-trust check (`get_trusted_proxy_url()` at line 238) now runs before `_base_url()`'s host validation (triggered here). If both `POSTHOG_HOST` is misconfigured and an ambient proxy is untrusted, the user now sees the less-actionable proxy-trust error first instead of the host-configuration error.

Blocking: no.

File: src/xagent/web/tools/mcp/posthog.py [hunk 8]
```diff
@@ -190,23 +191,64 @@ def _extract_error_detail(response: requests.Response) -> str | None:
     return detail if isinstance(detail, str) and detail else None
 
 
+def _make_request(**kwargs: Any) -> requests.Response:
```
Comment (rogercloud) [inline src/xagent/web/tools/mcp/posthog.py:194]: **Minor, non-blocking.** `_make_request(**kwargs: Any)` is fully untyped and its name is easily confused with the sibling `_request` function. No named/typed keyword params means a typo'd kwarg (e.g. `header=` for `headers=`) won't be caught by mypy.

Suggestion: named parameters, or a more distinct name like `_send_request`.

Blocking: no.

File: src/xagent/web/tools/mcp/posthog.py [hunk 9]
```diff
@@ -190,23 +191,64 @@ def _extract_error_detail(response: requests.Response) -> str | None:
...
+    # The result is then passed to `_make_request()` (see above), which
+    # issues the call on a trust_env=False Session -- disabling every
+    # ambient/OS-native proxy source, not just the env vars this function
+    # checks -- so "no proxy" here is an actual guarantee for this call.
+    proxy_url = get_trusted_proxy_url()
+    proxies = {"http": proxy_url, "https": proxy_url} if proxy_url else {}
+
     url = f"{_base_url()}{path}"
     try:
         for attempt in (0, 1):
```
Comment (rogercloud) [inline src/xagent/web/tools/mcp/posthog.py:261]: **Minor, non-blocking.** A single mutable `proxies` dict (built at line 239) is passed by reference to both `_make_request` calls across the two retry attempts in this loop. Currently harmless since `trust_env=False` and `allow_redirects=False` prevent in-place mutation between attempts, but would become a real hazard if either were changed later without revisiting this.

Blocking: no.

File: src/xagent/web/tools/mcp/posthog.py [hunk 10]
```diff
@@ -9,6 +9,7 @@
 import requests
 from mcp.server.fastmcp import FastMCP
 
+from ....core.tools.core.web_content import get_trusted_proxy_url
```
Comment (rogercloud) [inline src/xagent/web/tools/mcp/posthog.py:12]: **Minor, non-blocking.** This `web/tools/mcp/` connector reaches into `core/tools/core/web_content.py` (a web-scraping module) for one helper, pulling in transitive deps (html2text, httpx, bs4). Precedent exists (`download_web_asset.py` does the same cross-package import). Not a defect, just a code-organization observation -- see the review body's finding 12 for a suggested relocation of `get_trusted_proxy_url()` to `core/utils/security.py`.

Blocking: no.

File: tests/web/tools/test_posthog_mcp.py [hunk 11]
```diff
@@ -346,11 +346,92 @@ def test_request_uses_configured_host_and_headers(monkeypatch):
...
+    # get_environ_proxies() -> urllib.request.getproxies(), which itself
+    # falls back *past* HTTP_PROXY/HTTPS_PROXY/ALL_PROXY to the OS's own
+    # proxy configuration (getproxies_macosx_sysconf() on macOS,
+    # getproxies_registry() on Windows) once none of those env vars are
+    # set. trust_env=False skips that entire environment-and-OS-native
+    # lookup in one place, so assert _make_request actually sets it before
+    # issuing the real request.
+    captured = {}
+
+    class FakeSession:
```
Comment (rogercloud) [inline tests/web/tools/test_posthog_mcp.py]: **Minor, non-blocking.** Missing test coverage for three specific paths: (a) `FakeSession.__exit__` being called is never asserted here. (b) No test for an `HTTP_PROXY`-only environment (no `HTTPS_PROXY`): `get_proxy_url()` collapses to a single value fanned into both the `http` and `https` proxy-dict keys (`posthog.py:239`), so `HTTP_PROXY` alone now also proxies HTTPS traffic, untested here (matches an existing codebase-wide convention, e.g. `linkedin.py:220-221`, so not posthog-specific). (c) `get_proxy_url()` doesn't read `ALL_PROXY`/`all_proxy` in any case variant -- not a regression, just an existing gap this PR's tests don't probe.

Blocking: no.

**Review by rogercloud — APPROVED** [review]
## Summary

This PR closes an SSRF bypass specific to the PostHog MCP connector: previously `requests.request()` ran with default `trust_env=True`, so an ambient proxy (an `HTTP(S)_PROXY` env var, or OS-native proxy config that this same module's `setup_proxy_env()` promotes into env at import) could perform DNS resolution on the proxy's side, silently evading the connector's own private-network hostname validation. The fix adds a `get_trusted_proxy_url()` gate (raises unless `XAGENT_TRUSTED_EGRESS_PROXY=1` is explicitly set when a proxy is present) and a `_make_request()` helper that issues the actual HTTP call on a `trust_env=False` `requests.Session`, so no ambient/OS-native proxy source is honored unless explicitly trusted. Tests are expanded accordingly.

**Blocking: no — recommended event: APPROVE**

## Design verdict

The security reasoning is sound and was independently verified against the installed `requests` 2.34.2 source: a `trust_env=True` session with `proxies={}` still injects environment/OS-native proxies via `merge_environment_settings`, so passing `proxies=` alone would **not** have closed the hole — `trust_env=False` is genuinely required, and it's implemented correctly. This is a real fix, not a patch over the symptom. A few design-level reservations remain (duplication with a sibling connector, and a CA-bundle reasoning gap), captured as non-blocking findings below.

## Findings

### Major

1. **Test coverage gap: the mock is too shallow to catch a regression in the exact security guards this PR adds.**
   `tests/web/tools/test_posthog_mcp.py:401-427` — `FakeSession.request(self, **kwargs)` only captures `trust_env`; it ignores every other kwarg. The only test exercising the real `_make_request` body (`test_make_request_disables_trust_env`) only asserts `trust_env is False`. Every other test that asserts on `proxies=`/`allow_redirects=` mocks `_make_request` itself away, so its real body is never exercised elsewhere. Mutation-testing reasoning: if `_make_request`'s body silently dropped `proxies=`/`allow_redirects=False` before calling `session.request()` — stripping the SSRF-egress gate and the redirect/credential-leak guard this PR exists to add — the full suite would still pass.
   Suggestion: make `FakeSession.request` capture the full `kwargs` dict and assert `proxies`/`allow_redirects` in `test_make_request_disables_trust_env`.
   Blocking: no.

### Medium

2. **CA bundle docstring reasoning doesn't hold for the deployment scenario this PR enables.**
   `src/xagent/web/tools/mcp/posthog.py:206-211` — the docstring argues no CA-bundle re-application is needed because "PostHog Cloud's two hosts always terminate TLS with a publicly trusted CA." But `trust_env=False` also disables `requests`' `REQUESTS_CA_BUNDLE`/`CURL_CA_BUNDLE` lookup (gated behind the same `if self.trust_env:` check, verified against the installed requests source), and the only scenario where a proxy is actually used here is when an operator sets `XAGENT_TRUSTED_EGRESS_PROXY=1` — a corporate egress proxy, the textbook case for TLS interception with an internally-issued CA, not a publicly trusted one. Concrete failure: an operator behind a TLS-intercepting corporate proxy who opts in expects `REQUESTS_CA_BUNDLE` to be honored (as with most other `requests`-based tools) but instead gets `SSLError: certificate verify failed` on every call with no way to fix it. It fails closed (no security hole), but the "trusted proxy" feature is non-functional in the deployment most likely to need it.
   Suggestion: explicitly apply `REQUESTS_CA_BUNDLE`/`CURL_CA_BUNDLE` to `verify=` when `trust_env=False`, or correct the docstring to state this limitation honestly.
   Blocking: no.

### Minor

3. **`NO_PROXY` is no longer honored — a genuine regression vs. pre-PR behavior.**
   `src/xagent/web/tools/mcp/posthog.py:239` — `NO_PROXY` bypass logic only runs inside `merge_environment_settings`'s `if self.trust_env:` branch; with `trust_env=False` and an explicitly-passed `proxies=` dict, it's never consulted. Pre-PR, `requests.request()` was called with no explicit `proxies=` and default `trust_env=True`, so `NO_PROXY=us.posthog.com` was respected. Post-PR it isn't once `XAGENT_TRUSTED_EGRESS_PROXY=1` is set. Direction is fail-secure (more traffic goes through the operator's already-trusted proxy, not less), so this is an operational/behavior-change note, not a security issue.
   Blocking: no.

4. **The documented `XAGENT_TRUSTED_EGRESS_PROXY=1` remedy isn't reachable via the self-service connect flow; only an admin-level global config edit works, and the error message doesn't say so.**
   `src/xagent/web/tools/mcp/posthog.py:238` (error text) and `src/xagent/web/api/mcp.py` (self-service `/apps/{app_id}/connect` endpoint, `required_env = ["POSTHOG_API_KEY", "POSTHOG_HOST"]` filters submitted env). The PostHog connector runs as its own stdio subprocess whose env comes from `get_default_environment()` plus `server.env`/`resolve_stdio_env(...)`, and the self-service connect path only forwards the two required PostHog keys — never `XAGENT_TRUSTED_EGRESS_PROXY`. An admin can set it via `PUT /servers/{id}` (applies `config.env` without the `required_env` filter), so there is a working remedy — but the error message the user sees doesn't distinguish "set it as a host/shell env var" (won't work for this connector) from "have an admin add it to the connector's global MCP server config" (does work).
   Suggestion: clarify the error message/docs to point at the actual admin remedy.
   Blocking: no.

5. **Three code comments reference `magento.py`, a file not present in this PR's branch or on `main`.**
   `src/xagent/web/tools/mcp/posthog.py:204, 209, 229` — `magento.py` exists on a separate unmerged sibling branch by the same author with a near-identical `_make_request`, so the comparison is technically accurate today, but a reader of `main` after this PR merges (before, or unless, that sibling PR also lands) will hit a comment pointing at a file they cannot find, with no cross-reference in this PR's description.
   Suggestion: reference the sibling PR number explicitly, or describe the rationale standalone without naming a not-yet-merged file.
   Blocking: no.

6. **Stale comment describes a scenario no longer reachable after this PR's change.**
   `src/xagent/web/tools/mcp/posthog.py:270-274` — inside `except requests.RequestException`, the comment still says a `ProxyError` could echo "the ambient HTTPS_PROXY URL... setup_proxy_env() exports whatever the OS has configured," but post-PR `_make_request` is only ever called with a proxy that already passed `get_trusted_proxy_url()`'s trust check, or no proxy at all — an untrusted ambient proxy now raises before reaching this code. The corresponding test comment (`tests/web/tools/test_posthog_mcp.py:493-495`) was updated to say "connecting through a trusted proxy"; this source comment was missed.
   Blocking: no.

7. **Proxy-trust check now runs before host validation, changing which error surfaces first.**
   `src/xagent/web/tools/mcp/posthog.py:238` vs `:241` — `get_trusted_proxy_url()` (line 238) now runs before `_base_url()`'s host validation (triggered at line 241). If both `POSTHOG_HOST` is misconfigured and an ambient proxy is untrusted, the user now sees the less-actionable proxy-trust error first instead of the host-configuration error.
   Blocking: no.

8. **`_make_request(**kwargs: Any)` is fully untyped and its name is easily confused with the sibling `_request` function.**
   `src/xagent/web/tools/mcp/posthog.py:194` (definition), `:244` (call site). No named/typed keyword params means a typo'd kwarg (e.g. `header=` for `headers=`) won't be caught by mypy.
   Suggestion: named parameters, or a more distinct name like `_send_request`.
   Blocking: no.

9. **A single mutable `proxies` dict is reused across both retry attempts (latent hazard only).**
   `src/xagent/web/tools/mcp/posthog.py:239, 243` — the `proxies` dict is built once outside the `for attempt in (0, 1):` loop and passed by reference to both `_make_request` calls. Currently harmless because `trust_env=False` and `allow_redirects=False` prevent in-place mutation between attempts, but would become a real hazard if either were changed later without revisiting this.
   Blocking: no.

10. **Layering smell: a `web/tools/mcp/` connector reaches into `core/tools/core/web_content.py` (a web-scraping module) for one helper, pulling in transitive deps (html2text, httpx, bs4).**
    `src/xagent/web/tools/mcp/posthog.py:12` — precedent exists (`download_web_asset.py` does the same cross-package import). Not a defect, just a code-organization observation — see finding 12 for a suggested relocation.
    Blocking: no.

11. **Missing test coverage for three specific paths.**
    `tests/web/tools/test_posthog_mcp.py` — (a) `FakeSession.__exit__` being called is never asserted (`:401-411`). (b) No test for an `HTTP_PROXY`-only environment (no `HTTPS_PROXY`): `get_proxy_url()` collapses to a single value fanned into both the `http` and `https` proxy-dict keys (`posthog.py:239`), so `HTTP_PROXY` alone now also proxies HTTPS traffic, untested here (matches an existing codebase-wide convention, e.g. `linkedin.py:220-221`, so not a posthog-specific bug). (c) `get_proxy_url()` doesn't read `ALL_PROXY`/`all_proxy` in any case variant — not a regression, just an existing gap this PR's tests don't probe.
    Blocking: no.

12. **Optional design nit: `get_trusted_proxy_url()` would fit better in `core/utils/security.py`.**
    `src/xagent/core/tools/core/web_content.py:106-122` (current location) vs. `src/xagent/core/utils/security.py`, which already hosts `PrivateNetworkHostError`, `reject_private_network_host`, and other egress-policy primitives that `posthog.py` already imports separately. Moving it would collapse two import lines into one in `posthog.py` rather than adding a dependency, and would need updating 3 call sites (`vision_tool.py`, `download_web_asset.py`, `posthog.py`). Purely optional.
    Blocking: no.

## Simplification opportunities

- `L194: shrink _make_request in posthog.py is 100% generic (no PostHog-specific logic) and duplicates a near-identical implementation already written for a sibling unmerged Magento connector PR. Extract it into src/xagent/web/tools/mcp/utils.py (already home to setup_proxy_env(), the function that manufactures this proxy-trust risk) so it can be reused — 28 of the ~30 other connector modules in this directory call setup_proxy_env() at import and have the identical latent SSRF-via-ambient-proxy gap this PR closes only for posthog.`

net: -0 lines now, but avoids a near-term second/third copy as this pattern spreads to other connectors

## Blocking status & recommended decision

Blocking: no. Findings 1-12 above are all non-blocking (test-coverage gap, CA-bundle docstring gap, `NO_PROXY` regression, unreachable self-service remedy, stale/dangling comments, error-ordering change, typing/naming nit, latent mutable-default hazard, layering nit, missing edge-case tests, optional relocation). No finding meets the blocking bar (incorrect user-visible result / contract violation / data loss / security bypass / severe perf regression under a credible supported-scenario trigger). The core design approach is sound and correctly closes the SSRF bypass it targets.

Recommended event: **APPROVE**

File: src/xagent/web/tools/mcp/posthog.py (hunk 3 above)
Comment (yiboyasss) [reply src/xagent/web/tools/mcp/posthog.py:257]: Agreed this is a real behavior change, but leaving it as-is for this PR: it only bites when an operator has *both* opted into `XAGENT_TRUSTED_EGRESS_PROXY=1` *and* set `NO_PROXY` for a posthog.com host specifically, the direction is fail-secure (more traffic through an already-trusted proxy, not less), and the sibling connector this PR's `_make_request` mirrors has the identical gap unfixed. Re-implementing `should_bypass_proxies`-equivalent logic on top of `trust_env=False` felt like scope creep for a fix targeted at the SSRF bypass. Not opening a tracking issue for this alone -- if it's worth doing, it's worth doing once for the shared pattern, not per-connector (see my reply below on the systemic angle).

File: src/xagent/web/tools/mcp/posthog.py (hunk 4 above)
Comment (yiboyasss) [reply src/xagent/web/tools/mcp/posthog.py]: Good catch, but this is a platform-level gap (the `/apps/{app_id}/connect` self-service flow's `required_env` filtering), not something specific to posthog.py or fixable within this PR's diff. I'll flag it to the maintainer for a decision on whether it's worth a tracking issue.

File: src/xagent/web/tools/mcp/posthog.py (hunk 7 above)
Comment (yiboyasss) [reply src/xagent/web/tools/mcp/posthog.py:259]: This matches the same ordering already shipped in the sibling connector's `_request()` (proxy-trust gate before host resolution) -- keeping it consistent there rather than diverging just for posthog. Agreed it's a minor UX rough edge, not fixing here.

File: src/xagent/web/tools/mcp/posthog.py (hunk 8 above)
Comment (yiboyasss) [reply src/xagent/web/tools/mcp/posthog.py:194]: Leaving `**kwargs: Any` as-is to stay consistent with the identical helper in the sibling connector fix this mirrors -- named/typed params would need to duplicate requests' own call signature to actually catch a typo'd kwarg, which felt like more surface area than the benefit justified. Open to revisiting if a typo'd kwarg actually bites someone.

File: src/xagent/web/tools/mcp/posthog.py (hunk 9 above)
Comment (yiboyasss) [reply src/xagent/web/tools/mcp/posthog.py:261]: Confirmed harmless today (trust_env=False + allow_redirects=False mean neither retry attempt can mutate the dict in place), and not fixing preemptively -- would rather revisit this if/when either of those actually changes, per your own note.

File: src/xagent/web/tools/mcp/posthog.py (hunk 10 above)
Comment (yiboyasss) [reply src/xagent/web/tools/mcp/posthog.py:12]: Agreed this isn't a defect (precedent already exists in download_web_asset.py), and the relocation you suggest in finding 12 is a reasonable idea, but it's a cross-cutting change (3+ call sites outside this PR's diff) I don't want to bundle into an SSRF-bypass fix. Flagging it alongside the systemic point below for a maintainer call on whether it's worth its own follow-up.

File: tests/web/tools/test_posthog_mcp.py (hunk 11 above)
Comment (yiboyasss) [reply tests/web/tools/test_posthog_mcp.py]: (a) is now covered -- I added a __exit__ assertion to test_make_request_disables_trust_env in the latest commit. (b) and (c) I'm leaving alone: as you noted, both are pre-existing codebase-wide gaps (matches linkedin.py's HTTP_PROXY/HTTPS_PROXY handling, and get_proxy_url() not reading ALL_PROXY predates this PR entirely), not something this diff introduces or narrows.

**Comment by yiboyasss** [comment]
Pushed a follow-up commit addressing the review:
- Fixed: re-applies `REQUESTS_CA_BUNDLE`/`CURL_CA_BUNDLE` under `trust_env=False` (would otherwise silently break the trusted-proxy path for the exact corporate-TLS-interception case it's meant for), strengthens `test_make_request_disables_trust_env` to assert the full kwargs (not just `trust_env`) reach `session.request()`, adds a CA-bundle test, and removes comments referencing an unmerged sibling file.
- Replied inline on the rest (`NO_PROXY` handling, the self-service connect flow's env-var reach, error ordering, `_make_request` typing, the mutable `proxies` dict, and the cross-package import) with reasoning for leaving them as-is in this PR.
